### ISSUE TITLE
How to speed up Home Assistant boot time when using this integration?

### ISSUE THREAD

**Issue by bertybuttface** [kongo09/philips-airpurifier-coap#76]
Hello,

I've found when booting Home Assistant this integration is always the last to load and slows down the rest of the boot process. I know this is probably quite a minor issue but is there anything we can do about this?

Cheers

**Comment by kongo09** [comment]
To be honest, I have no idea...

**Comment by TheBartoMan** [comment]
I second this. Its the one component that sits and takes forever at startup

**Comment by kongo09** [comment]
Any support in debugging this would be appreciated

**Comment by bertybuttface** [comment]
Does anyone know how to profile a home assistant extension?

**Comment by bertybuttface** [comment]
We can use this but we need to work out how to make it run on startup: https://www.home-assistant.io/integrations/profiler/

**Comment by Anashost** [comment]
I found two kinda dumb work arounds, pick one of them:
* disable this integration before restarting HA, then enable it back after restart is completed. This way HA will reboot normally and doesn't wait for this integration to start.

* once HA begins to start and hangs trying to start this integration.. if you open the Air+ app and do any command (i personally toggle the beep sound) then the integration will start right away, and HA complete loading the other stuff.

@kongo09 i don't know if this behavior can give you a clue on why this issue could happen in first place.

But to me it looks like the integration wait for a response from the device, which takes a lot of time to happen in most cases.

**Comment by kongo09** [comment]
Maybe there is a way to be more efficient with the initial data requests. If someone could look at that, I'd appreciate it.

**Comment by tjorim** [comment]
From @bdraco [here](https://discord.com/channels/330944238910963714/427516175237382144/1212811740350513172):
its doing some pretty obvious blocking I/O in the event loop https://github.com/kongo09/philips-airpurifier-coap/blob/90f8f0025d76d0c7891e0d9aaf70729c3780a4cc/custom_components/philips_airpurifier_coap/__init__.py#L72

**Comment by julianjwong** [comment]
> I found two kinda dumb work arounds, pick one of them:
> 
> * disable this integration before restarting HA, then enable it back after restart is completed. This way HA will reboot normally and doesn't wait for this integration to start.
> * once HA begins to start and hangs trying to start this integration.. if you open the Air+ app and do any command (i personally toggle the beep sound) then the integration will start right away, and HA complete loading the other stuff.
> 
> @kongo09 i don't know if this behavior can give you a clue on why this issue could happen in first place.
> 
> But to me it looks like the integration wait for a response from the device, which takes a lot of time to happen in most cases.

Is there a way to automate disabling of the integration prior to a restart and then reenable after startup?

**Comment by Flappiee** [comment]
I'm also having issues with startup of Home Assistant.
It says the "Air Purifier" addon is loading and stays there for a long time.

Don't know why. If i can provide any logging, please let me know what you need.

**Comment by chrisp250** [comment]
Same here. Workaround is to open the Philips app on my phone as HA tries to connect to the purifier during start up

**Comment by william-niq** [comment]
> Same here. Workaround is to open the Philips app on my phone as HA tries to connect to the purifier during start up

This is probably the same reason why this is in the readme:

>  timer code to attempt reconnects if the device is not responsive. Sometimes, that helps. But not always. You've been warned.

The app does not always fetch the status of the device like HA would. Also if you open the app you see a "connecting" prompt. This would indicate the app does some kind of api call to get the device out of sleep or something. 

Found this: 
```
// Empty message used for sending no data, but instead for getting the response
// In the legacy version this returns a single byte error code/state, and was called BasicStatusCheck
```
So at start up of the app you send a empty message just to get the device status before you the the actual data of the device. So at startup of the integration you could do something similar "checkDeviceProvision" or something.

**Comment by daandav** [comment]
Create an automation that turns on the purifyer just at reboot event of HA. Turn off when finished booting. Works like a charm.

**Comment by jubie25** [comment]
> Create an automation that turns on the purifyer just at reboot event of HA. Turn off when finished booting. Works like a charm.

Do you have some hints for us how you did this? Saves lot of time for the more unexperienced users .-)

**Comment by diplix** [comment]
i added a helper entity to keep track whether the purifier was on before the HA restart. you can leave that out of course. 

```
alias: philips air purifier start / stop at HA restart
description: ""
mode: single
triggers:
  - event: start
    id: up
    trigger: homeassistant
  - event: shutdown
    id: down
    trigger: homeassistant
conditions: []
actions:
  - choose:
      - conditions:
          - condition: trigger
            id:
              - down
        sequence:
          - action: fan.turn_on
            metadata: {}
            data: {}
            target:
              entity_id: fan.luftreiniger
          - if:
              - condition: state
                entity_id: fan.luftreiniger
                state: "on"
            then:
              - action: input_boolean.turn_on
                metadata: {}
                data: {}
                target:
                  entity_id: input_boolean.luftreiniger_lief_vor_ha_restart
            else:
              - action: input_boolean.turn_off
                metadata: {}
                data: {}
                target:
                  entity_id: input_boolean.luftreiniger_lief_vor_ha_restart
      - conditions:
          - condition: trigger
            id:
              - up
        sequence:
          - if:
              - condition: state
                entity_id: input_boolean.luftreiniger_lief_vor_ha_restart
                state: "on"
            then:
              - action: fan.turn_off
                metadata: {}
                data: {}
                target:
                  entity_id: fan.luftreiniger
          - action: input_boolean.turn_off
            metadata: {}
            data: {}
            target:
              entity_id: input_boolean.luftreiniger_lief_vor_ha_restart
```

**Comment by kongo09** [comment]
@eulemitkeule worked on a fork here https://github.com/EuleMitKeule/philips-airpurifier-coap that refactors parts of the integration and seems to improve it. Maybe I'll get a PR or I'll find time to pull the stuff myself.

**Comment by kongo09** [comment]
see #178

**Comment by kongo09** [comment]
closed by #186
